### PR TITLE
BUG: stats: use mean behavior for percentileofscore in bootstrap

### LINKED ISSUE #15883: BUG: stats.bootstrap bca implementation triggers ValueError for `min` statistic
### Describe your issue.

`stats.bootstrap` uses a custom implementation of `percentileofscore` that exclusively uses the `strict` behavior from the primary implementation.

As a result, the `min` statistic is always identified as the zeroth percentile, for all datasets. The bca then chooses -inf for z0_hat, which causes a ValueError when the lower percentile is selected as -inf.

I recognize that using bootstrap to estimate confidence intervals of `min` and `max` is not really intended behavior. However, I believe the behavior should be consistent for both of those estimators, and that the `mean` behavior of `percentileofscore` should be used instead.

### Reproducing Code Example

```python
import scipy.stats as stats
import numpy as np
norm = stats.norm.rvs(0, 1, 100)
stats.bootstrap((norm,), statistic=np.min)
```

### Error message

```shell
ValueError                                Traceback (most recent call last)
Input In [1], in <cell line: 4>()
      2 import numpy as np
      3 norm = stats.norm.rvs(0, 1, 100)
----> 4 stats.bootstrap((norm,), statistic=np.min)

File ~/miniconda3/envs/analytics/lib/python3.9/site-packages/scipy/stats/_bootstrap.py:454, in bootstrap(data, statistic, vectorized, paired, axis, confidence_level, n_resamples, batch, method, random_state)
    451         return np.percentile(a=a, q=q, axis=-1)
    453 # Calculate confidence interval of statistic
--> 454 ci_l = percentile_fun(theta_hat_b, interval[0]*100)
    455 ci_u = percentile_fun(theta_hat_b, interval[1]*100)
    456 if method == 'basic':  # see [3]

File ~/miniconda3/envs/analytics/lib/python3.9/site-packages/scipy/stats/_bootstrap.py:78, in _percentile_along_axis(theta_hat_b, alpha)
     76 for indices, alpha_i in np.ndenumerate(alpha):
     77     theta_hat_b_i = theta_hat_b[indices]
---> 78     percentiles[indices] = np.percentile(theta_hat_b_i, alpha_i)
     79 return percentiles[()]

File <__array_function__ internals>:5, in percentile(*args, **kwargs)

File ~/miniconda3/envs/analytics/lib/python3.9/site-packages/numpy/lib/function_base.py:3866, in percentile(a, q, axis, out, overwrite_input, interpolation, keepdims)
   3864 q = asanyarray(q)  # undo any decay that the ufunc performed (see gh-13105)
   3865 if not _quantile_is_valid(q):
-> 3866     raise ValueError("Percentiles must be in the range [0, 100]")
   3867 return _quantile_unchecked(
   3868     a, q, axis, out, overwrite_input, interpolation, keepdims)

ValueError: Percentiles must be in the range [0, 100]
```

### SciPy/NumPy/Python version information

1.7.3 1.21.2 sys.version_info(major=3, minor=9, micro=7, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by jcampbell**
#### Reference issue
Closes gh-15883.

#### What does this implement/fix?
This updates the behavior of the percentileofscore implementation used by bootstrap to use the 'mean' behavior for calculating percentile, which avoids the 'strict' behavior of assigning the min value the zeroth percentile.

### CODE REVIEW

File: scipy/stats/_resampling.py [hunk 1]
```diff
@@ -80,12 +80,13 @@ def _bootstrap_resample(sample, n_resamples=None, random_state=None):
...
 def _percentile_of_score(a, score, axis):
     """Vectorized, simplified `scipy.stats.percentileofscore`.
+    Uses logic of the 'mean' value of percentileofscore's kind parameter.
 
     Unlike `stats.percentileofscore`, the percentile returned is a fraction
     in [0, 1].
     """
     B = a.shape[axis]
-    return (a < score).sum(axis=axis) / B
+    return ((a < score).sum(axis=axis) + (a <= score).sum(axis=axis)) / (2 * B)
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:89]: Oops, I misread how `kind='weak'` worked initially. This is great. Very simple.

**Comment by mdhaber** [comment]
I allowed tests to run. I'll take a look at the results tomorrow and play with this a bit, but looks good!

**Comment by mdhaber** [comment]
Test failure is in `main` (unrelated).

File: scipy/stats/tests/test_resampling.py [hunk 2]
```diff
@@ -388,6 +388,22 @@ def test_bootstrap_gh15678(method):
...
+    dist = stats.norm(loc=2, scale=4)
+    data = dist.rvs(size=100, random_state=rng)
+    true_min = np.min(data)
+    data = (data,)
+    res = bootstrap(data, np.min, method="BCa", n_resamples=100,
+                    random_state=np.random.default_rng(3942))
+    assert true_min == res.confidence_interval.low
+    assert isinstance(res.confidence_interval.low, np.float64)
+    assert isinstance(res.confidence_interval.high, np.float64)
+    assert isinstance(res.standard_error, np.float64)
```
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:404]: I don't think the output type is really appropriate for this test. 
But we can do something stronger based on what you wrote in the original issue:
```suggestion
    res2 = bootstrap(-np.array(data), np.max, method="BCa", n_resamples=100,
                     random_state=np.random.default_rng(3942))
    assert_allclose(-res.confidence_interval.low, res2.confidence_interval.high)
    assert_allclose(-res.confidence_interval.high, res2.confidence_interval.low)
```

Comment (mdhaber) [reply scipy/stats/tests/test_resampling.py]: @jcampbell I took the liberty of making this change. Look good to you? If so, and if tests pass, I'll merge.

File: scipy/stats/tests/test_resampling.py [hunk 3]
```diff
@@ -388,6 +388,23 @@ def test_bootstrap_gh15678(method):
...
+    data = dist.rvs(size=100, random_state=rng)
+    true_min = np.min(data)
+    data = (data,)
+    res = bootstrap(data, np.min, method="BCa", n_resamples=100,
+                    random_state=np.random.default_rng(3942))
+    assert true_min == res.confidence_interval.low
+    res2 = bootstrap(-np.array(data), np.max, method="BCa", n_resamples=100,
+                     random_state=np.random.default_rng(3942))
+    assert_allclose(-res.confidence_interval.low, res2.confidence_interval.high)
+    assert_allclose(-res.confidence_interval.high, res2.confidence_interval.low)
```
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:405]:
```suggestion
    assert_allclose(-res.confidence_interval.low,
                    res2.confidence_interval.high)
    assert_allclose(-res.confidence_interval.high,
                    res2.confidence_interval.low)
```

File: scipy/stats/tests/test_resampling.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/tests/test_resampling.py]: All test passed except the `TestMoment` issue in `main` and a PEP8 issue, which I fixed, so I didn't re-run CI.

**Comment by jcampbell** [comment]
Excellent -- thank you for improving the test, and very excited to have this merged!

**Comment by mdhaber** [comment]
Thanks @jcampbell! I was all set to merge this, but I thought I'd refer to the source I used when writing `bootstrap`, Efron and Tibshirani's Introduction to the Bootstrap. They explicitly define the percentile of score as the "percentage of bootstrap replications _less than_ the original estimate":

image

Compare against code:
https://github.com/scipy/scipy/blob/9a12908f843aed87203b32e45d1001353d90c548/scipy/stats/_resampling.py#L115-L118

So that's why I wrote `_percentile_of_score` with strict less than, originally.

Now, this PR seems to me to improve the BCa bootstrap. But I'm uncertain of why the authors wrote this with strict inequality- 
 are there deep reasons, or was it was just an oversight not expected to make a difference in practice?

Do you have any other sources that could clarify? If not, I should probably get a second opinion before merging this, even though I agree from an intuitive perspective.

**Comment by jcampbell** [comment]
@mdhaber thanks for looking into it more -- I'm with you in that I don't know whether there are deeper reasons that I'm not seeing, and your implementation is certainly exactly what's written in the referenced book. I'd like to look into it a bit more before closing/revising the PR if that's okay, because now I'm quite curious!

**Comment by mdhaber** [comment]
FInd anything, @jcampbell?

**Comment by mdhaber** [comment]
@tupui @tirthasheshpatel @chrisb83 @ev-br can you think of anyone that might be able to answer this question for us?
If not, would you support this change? 
It is pretty straightforward, it only affects edge cases, and it seems desirable. (See how the new unit test confirms the symmetry w.r.t. `min` and `max` statistics? In `main`, this symmetry is violated, and worse, there would be an error!) But I'm hesitant to go against the reference without more eyes on it.

**Comment by tupui** [comment]
You can try sending an email to Art Owen (can share his contact with you if you don't have it).

**Comment by josef-pkt** [comment]
Not directly related to the bootstrap. I haven't looked at details for it in a long time.

for continuous random variables p(y* < y) is the same as p(y*<= y). So often references don't make a clear distinction.

However, for discrete random variables like binomial or poisson counts, there are ties with positive probability.
"exact" p-values for those use "at least as extreme" i.e. `p(y*<= y)`
However, those "exact" tests for count variables are usually very conservative.

An adjustment is to use "midp" which is `p(y* < y) + 0.5 * p(y* = y)` which is similar to the change here AFAICS.
"midp" tests have on average the correct size but in contrast to exact they don't guarantee that the test does not overreject in all cases (i.e. is liberal for some parameters).

For binomial and poisson hypothesis tests and confidence intervals, I add an option to choose between "exact" and "midp" in statsmodels (similar to related R packages). 
If ties occur with positive probability, then which kind of inequality is used is not just a computational issue.

Aside: I don't think bootstrap is valid for the `min` statistic as in the original issue.

**Comment by chrisb83** [comment]
> can you think of anyone that might be able to answer this question for us?
Apologies, I cannot help.

> If not, would you support this change
I found the argument about the symmetry w.r.t. min/max convincing. But I do not know enough about the background to give you qualified feedback.

**Comment by jcampbell** [comment]
@mdhaber Thanks for continuing to push on this.

I'd be happy to reach out to Art Owen and/or participate in that discussion. It's definitely a subtle issue.

A couple synthesis thoughts -- I agree with @chrisb83 that the symmetry argument is a useful one. I wanted to look at a couple of simulations to help build intuition, which I'll try to run tonight.

**Comment by jcampbell** [comment]
I don't think simulations would provide a justification, but in this case what I see suggests this PR should be closed:

beta_plots

**Comment by mdhaber** [comment]
Hmm - could you interpret the results of the simulations? Do they just show that bootstrap is not good for `max` and `min`?

**Comment by jcampbell** [comment]
I was broadly observing that in the `min` and `max` examples, the intervals are smaller (hence less overlap and so lower peaks in the histogram), and the patched version lands closer to the `basic` method, suggesting that abandoning the attempt to correct for skewness may simply be the better approach when at those extreme statistics with bootstrap (subject to how you want to interpret that anyway).

More to the point: `bca` didn't explode, which leads me to ask for clarification for the parent issue --> has another change been made that no longer marks the distribution as degenerate, and now does return a confidence interval even for `min` and even for `bca`?

**Comment by mdhaber** [comment]
> has another change been made that no longer marks the distribution as degenerate, and now does return a confidence interval even for min and even for bca?

No. This PR wasn't merged, so [this](https://github.com/scipy/scipy/issues/15883#issuecomment-1079775686) is still the behavior for your original example. I don't know how the plots were generated, so I'm not sure why it didn't have a problem for the `min` statistic here.

**Comment by jcampbell** [comment]
Thank you for confirming that no other change had been merged. I think I actually made a mistake in my environment and wasn't correctly comparing the patch.

I looked again and regenerated the data. I'm going to leave the thread intact, but will drop in a new chart here. I think the case made here is what I'd hoped when creating the patch in the first place: it makes it possible to support min without loss of capability visible in the simulation. I'll respond more on the original issue too. Thank you again!


20220608_bca_vs_bca_patch


PS -- there was a fun side project of looking at constructing the histograms from the confidence interval data...is there a (standard) library for that?

**Comment by mdhaber** [comment]
Are you using the same seed for all the simulations? (If not, could you do that? If so, I'm surprised that results aren't identical for the first four sample moments.)

> constructing the histograms from the confidence interval data

I'm not sure I understand. Could you elaborate?

**Comment by jcampbell** [comment]
No -- I wasn't fixing the seed. Thanks for pointing that out -- that does make the result much more clear, and in favor of the change.

output


(re: the intervals -- I meant that the histogram is constructed by taking the union of endpoints of the confidence intervals as the bin edges, and then counting how many confidence intervals included the resulting sub-divided intervals (rather than taking a value inside the interval and constructing bins using those datapoints). I wasn't sure if there's a standard practice for that.

```
def build_interval_histogram(intervals: np.array):
    shape = intervals.shape
    try:
        assert len(shape) == 2, "intervals must be a 2d array"
        assert shape[1] == 2, "each interval must have two endpoints"
        assert np.all(intervals[:,0] <= intervals[:,1]), "each interval must have min <= max"
    except AssertionError as e:
        raise ValueError(str(e))

    bins = np.unique(intervals)
    cnts = np.zeros(shape=len(bins) - 1, dtype=np.int64)

    for k in range(len(bins) - 1):
        for interval in intervals[:]:
            # Does any portion of this interval intersect with this bin
            if interval[0] < bins[k+1] and interval[1] > bins[k]:
                cnts[k] += 1

    return cnts, bins
```

**Comment by mdhaber** [comment]
I'm not aware of a function or library for that, nope. 

Can you send the rest of the code for this? Are all these intervals coming from resampling the same sample, or different samples drawn from the same distribution?

I noticed that for some of those plots toward the right, the total area within the blue region appears to be less than that in the orange region. Would that change if the interval comparison were not strict? That is, instead of
```python
if interval[0] < bins[k+1] and interval[1] > bins[k]:
```
would the areas be equal if it were
```python
if interval[0] <= bins[k+1] and interval[1] >= bins[k]:
```

**Comment by jcampbell** [comment]
@mdhaber -- here's the replication code, as well as a patch for _resampling.py corresponding to allowing the `bca_patch` method.

```
# %%
import numpy as np
from scipy import stats as stats
import matplotlib.pyplot as plt
import functools

# %%
a_s = np.linspace(1, 3, num=11)
b_s = np.linspace(4, 1, num=11)
params = zip(a_s, b_s)
betas = [stats.beta(*k) for k in params]

# %%
for beta in betas:
    x = np.linspace(beta.ppf(0.01),
                beta.ppf(0.99), 100)
    a, b = beta.args
    plt.plot(x, beta.pdf(x), alpha=0.6, label=f'beta({round(a, 2)}, {round(b, 2)}) pdf')
plt.title("Beta Distributions Used in Experiment")
plt.legend(bbox_to_anchor=(1., 1))
plt.show()

# %%
n_samples = 40
sample_size = 20
res = dict()

q_01 = functools.partial(np.quantile, q=0.01)
q_99 = functools.partial(np.quantile, q=0.99)

functions = {
    "mean": np.mean,
    "var": np.var,
    "skew": stats.skew,
    "kurt": stats.kurtosis,
    "min": np.min,
    "max": np.max,
    "q_01": q_01,
    "q_99": q_99
}
# methods = ["percentile", "basic", "bca_patch", "bca"]
methods = ["bca_patch", "bca"]


for beta_idx, beta in enumerate(betas):
    state = np.random.default_rng(2022061001)
    rvs = beta.rvs((n_samples, sample_size), random_state=state)
    for k in range(n_samples):
        res[beta_idx, k, "true_mean"], res[beta_idx, k, "true_var"], res[beta_idx, k, "true_skew"], res[beta_idx, k, "true_kurt"] = beta.stats('mvsk')
        res[beta_idx, k, "true_min"] = beta.ppf(0.)
        res[beta_idx, k, "true_max"] = beta.ppf(1.)
        res[beta_idx, k, "true_q_01"] = beta.ppf(0.01)
        res[beta_idx, k, "true_q_99"] = beta.ppf(0.99)
        for method in methods:
            for fn_name, fn in functions.items():
                state = np.random.default_rng(2022061002)
                res[beta_idx, k, fn_name, method] = stats.bootstrap((rvs[k],), statistic=fn, method=method, random_state=state)

# %%
def build_interval_histogram(intervals: np.array):
    shape = intervals.shape
    try:
        assert len(shape) == 2, "intervals must be a 2d array"
        assert shape[1] == 2, "each interval must have two endpoints"
        assert np.all(intervals[:,0] <= intervals[:,1]), "each interval must have min <= max"
    except AssertionError as e:
        raise ValueError(str(e))

    bins = np.unique(intervals)
    cnts = np.zeros(shape=len(bins) - 1, dtype=np.int64)

    for k in range(len(bins) - 1):
        for interval in intervals[:]:
            # Does any portion of this interval intersect with this bin
            if interval[0] < bins[k+1] and interval[1] > bins[k]:
                cnts[k] += 1

    return cnts, bins

# %%
fig = plt.figure(figsize=(60, 120)) # constrained_layout=True)
# fig.suptitle("Comparing Boostrap Methods: Skewed distributions and tail statistics")
subfigs = fig.subfigures(nrows=len(betas), ncols=1)

legend_labels = []

for beta_idx, subfig in enumerate(subfigs):
    a, b = betas[beta_idx].args
    subfig.suptitle(f'beta({round(a, 2)}, {round(b, 2)})')
    axs = subfig.subplots(nrows=1, ncols=len(functions))

    for idx, stat in enumerate(functions.keys()):
        legend = []
        for method in methods:
            hist, bins = None, None
            intervals = []
            for k in range(n_samples):
                intervals.append((
                    res[beta_idx, k, stat, method].confidence_interval.low,
                    res[beta_idx, k, stat, method].confidence_interval.high
                ))
            try:
                hist, bins = build_interval_histogram(np.array(intervals))
            except ValueError:
                hist, bins = None, None
            if bins is not None:
                b_obj = axs[idx].bar(bins[:-1], hist, width=np.diff(bins), align='edge', alpha=0.4)
                legend.append(b_obj)
                legend_labels.append(method)
        v_obj = axs[idx].axvline(res[beta_idx, k, f"true_{stat}"], color='r')
        legend.append(v_obj)
        legend_labels.append("true")
        axs[idx].legend(legend, legend_labels)
        axs[idx].set_title(stat)

```

```
diff --git a/scipy/stats/_resampling.py b/scipy/stats/_resampling.py
index d436e1c1a..f1e6181fb 100644
--- a/scipy/stats/_resampling.py
+++ b/scipy/stats/_resampling.py
@@ -88,6 +88,16 @@ def _percentile_of_score(a, score, axis):
     return (a < score).sum(axis=axis) / B
 
 
+def _percentile_of_score_patch(a, score, axis):
+    """Vectorized, simplified `scipy.stats.percentileofscore`.
+
+    Unlike `stats.percentileofscore`, the percentile returned is a fraction
+    in [0, 1].
+    """
+    B = a.shape[axis]
+    return ((a < score).sum(axis=axis) + (a <= score).sum(axis=axis)) / (2 * B)
+
+
 def _percentile_along_axis(theta_hat_b, alpha):
     """`np.percentile` with different percentile for each slice."""
     # the difference between _percentile_along_axis and np.percentile is that
@@ -107,6 +117,37 @@ def _percentile_along_axis(theta_hat_b, alpha):
     return percentiles[()]  # return scalar instead of 0d array
 
 
+def _bca_interval_patch(data, statistic, axis, alpha, theta_hat_b, batch):
+    """Bias-corrected and accelerated interval."""
+    # closely follows [2] "BCa Bootstrap CIs"
+    sample = data[0]  # only works with 1 sample statistics right now
+
+    # calculate z0_hat
+    theta_hat = np.asarray(statistic(sample, axis=axis))[..., None]
+    percentile = _percentile_of_score_patch(theta_hat_b, theta_hat, axis=-1)
+    z0_hat = ndtri(percentile)
+
+    # calculate a_hat
+    theta_hat_i = []  # would be better to fill pre-allocated array
+    for jackknife_sample in _jackknife_resample(sample, batch):
+        theta_hat_i.append(statistic(jackknife_sample, axis=-1))
+    theta_hat_i = np.concatenate(theta_hat_i, axis=-1)
+    theta_hat_dot = theta_hat_i.mean(axis=-1, keepdims=True)
+    num = ((theta_hat_dot - theta_hat_i)**3).sum(axis=-1)
+    den = 6*((theta_hat_dot - theta_hat_i)**2).sum(axis=-1)**(3/2)
+    a_hat = num / den
+
+    # calculate alpha_1, alpha_2
+    z_alpha = ndtri(alpha)
+    z_1alpha = -z_alpha
+    num1 = z0_hat + z_alpha
+    alpha_1 = ndtr(z0_hat + num1/(1 - a_hat*num1))
+    num2 = z0_hat + z_1alpha
+    alpha_2 = ndtr(z0_hat + num2/(1 - a_hat*num2))
+    return alpha_1, alpha_2
+
+
+
 def _bca_interval(data, statistic, axis, alpha, theta_hat_b, batch):
     """Bias-corrected and accelerated interval."""
     # closely follows [2] "BCa Bootstrap CIs"
@@ -201,7 +242,7 @@ def _bootstrap_iv(data, statistic, vectorized, paired, axis, confidence_level,
         if batch != batch_iv or batch_iv <= 0:
             raise ValueError("`batch` must be a positive integer or None.")
 
-    methods = {'percentile', 'basic', 'bca'}
+    methods = {'percentile', 'basic', 'bca', 'bca_patch'}
     method = method.lower()
     if method not in methods:
         raise ValueError(f"`method` must be in {methods}")
@@ -480,6 +521,10 @@ def bootstrap(data, statistic, *, n_resamples=9999, batch=None,
         interval = _bca_interval(data, statistic, axis=-1, alpha=alpha,
                                  theta_hat_b=theta_hat_b, batch=batch)
         percentile_fun = _percentile_along_axis
+    elif method == 'bca_patch':
+        interval = _bca_interval_patch(data, statistic, axis=-1, alpha=alpha,
+                                 theta_hat_b=theta_hat_b, batch=batch)
+        percentile_fun = _percentile_along_axis
     else:
         interval = alpha, 1-alpha
 
```

**Comment by mdhaber** [comment]
@Micky774 Thought you might be interested in this. The proposed change is simple and it resolves some problems. 

In main, the proposed test fails for two reasons - a warning would be generated, and the confidence interval for the `min` statistic would be `nan`. 

With the PR, we resolve the warning, and we get some symmetry between `min` and `max` statistics. 

Note that this change only affects cases in which there are ties in the bootstrap distribution of the statistic, so it would have been reasonable for Efron to neglect the distinction between `<` and `<=` in [Introduction to the Bootstrap](https://github.com/scipy/scipy/pull/15892#issuecomment-1083488771). (Incidentally, he also doesn't add one to the numerator and denominator in randomized hypothesis tests.)

Here is some code to think about...
```python3
import numpy as np
from numpy.testing import assert_allclose
from scipy import stats
from scipy.stats import bootstrap

rng = np.random.default_rng(1891289180021102)
dist = stats.norm(loc=2, scale=4)
data = dist.rvs(size=100, random_state=rng)
true_min = np.min(data)
data = (data,)
res = bootstrap(data, np.min, method="BCa", n_resamples=100,
                random_state=np.random.default_rng(3942))
res2 = bootstrap(-np.array(data), np.max, method="BCa", n_resamples=100,
                 random_state=np.random.default_rng(3942))
assert_allclose(-res.confidence_interval.low,
                res2.confidence_interval.high)
assert_allclose(-res.confidence_interval.high,
                res2.confidence_interval.low)
```

File: scipy/stats/tests/test_resampling.py [hunk 4]
```diff
@@ -470,6 +470,23 @@ def test_bootstrap_gh15678(method):
     assert_allclose(res.standard_error, ref.standard_error)
     assert isinstance(res.standard_error, np.float64)
 
+def test_bootstrap_min():
```
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:473]:
```suggestion

def test_bootstrap_min():
```

File: scipy/stats/tests/test_resampling.py [hunk 5]
```diff
@@ -470,6 +470,23 @@ def test_bootstrap_gh15678(method):
...
+    data = (data,)
+    res = bootstrap(data, np.min, method="BCa", n_resamples=100,
+                    random_state=np.random.default_rng(3942))
+    assert true_min == res.confidence_interval.low
+    res2 = bootstrap(-np.array(data), np.max, method="BCa", n_resamples=100,
+                     random_state=np.random.default_rng(3942))
+    assert_allclose(-res.confidence_interval.low,
+                    res2.confidence_interval.high)
+    assert_allclose(-res.confidence_interval.high,
+                    res2.confidence_interval.low)
```
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:489]:
```suggestion
                    res2.confidence_interval.low)

```

**Comment by mdhaber** [comment]
The macOS Test Matix failure is in `scipy/sparse/linalg/_eigen/arpack/tests/test_arpack.py`.
The linux_aarch64_test failure is in `scipy/linalg/tests/test_solvers.py::test_solve_discrete_are`.
The macos_arm64_test failure is a bunch of warnings in stats distributions. 
The lint failure will be fixed by the next commit.

Art Owen mentioned via email that one way of handling this is to "count each tie as 50%". That seems to be equivalent to what we're doing here: since we average the results with `<` and `<=`, every tie (in which strict equality is satisfied) is counted with a weight of 50%.

This way is "ad hoc"; the preferred way is to "make whatever choice gives the less statistically significant result". I'll go ahead and merge this PR and submit a PR for the preferred way shortly.

Thanks @jcampbell!

**Pull request merged**